pandas/core/groupby/ops.py

```diff
@@ -639,6 +639,10 @@ def agg_series(self, obj: Series, func: F):
 
         try:
             return self._aggregate_series_fast(obj, func)
+        except AssertionError:
+            # in some cases (eg GH-36003) an internal AssertionError can be
+            # raised if libreduction cannot handle this case
+            pass
```

Comment:
@jbrockmendel I know you won't like this ... but, libreduction is for some reason I didn't yet figure out raising the AssertionError, which a user should not see. So suppressing that here fixes it.

(and specifically for AssertionError (not other errors), that might actually be OK, since AssertionErrors are almost always an indication that something internally in pandas went wrong (internal bug), and not a user error that should be bubbled up)

Reply:
yeah don't like this, can you instead handle at a lower level?

Reply:
Took a further look, and I think this is inherent to the way that libreduction does the "sliding" series trick with replacing the underlying values. 

The problem with this specific case if that the first Series it creates actually gets inferred as datetime64[ns] dtype (that's normal `Series(..)` behaviour, the first date is not out of bounds), but then for the next group, the blocks values are replaced with the object dtype array, which means you end up with an invalid DatetimeBlock with object dtype array.

I am not sure there is a way around that, except by not using this libreduction trick. 
We could maybe check that the types are equal when setting the block values:

https://github.com/pandas-dev/pandas/blob/78d149897201cb55b403f6f84fc541a3d7aa86ad/pandas/_libs/reduction.pyx#L55

but that might also undo the performance gain of this trick.

Reply:
what you're describing sounds similar to what motivated the libreduction edits in #35417.  would porting those make a difference?

Reply:
Could you try it out on that branch? Assuming I need to only look at the diff in reduction.pyx, I don't directly see how that would fix this issue, as it doesn't change how the Series' blocks' values are set within the loop?

Reply:
nope, doesn't fix it.

it looks like we're getting to `DatetimeBlock.array_values` with `self.values` of `array([datetime.datetime(3005, 1, 1, 10, 30, 23, 540000)], dtype=object)`.

Reply:
OK, a fix that does appear to work: in libreduction when defining `self.cached_typ`, pass `dtype=self.buf.dtype` to `self.typ`

Reply:
Ah, good catch. Will push that.